{"cells":[{"cell_type":"markdown","metadata":{"id":"_SmTogjeMD47"},"source":["# 생물정보학 및 실습 1 - Term Project용 환경 셋업\n","\n","```\n","생물정보학 및 실습 1\n","서울대학교 생물정보학전공/생명과학부 2024년 1학기\n","\n","이 노트북은 Creative Commons BY-SA 조건으로 재사용 가능합니다.\n","장혜식 <hyeshik@snu.ac.kr>, 서울대학교 생명과학부\n","```\n","\n","이 노트북은 Google Colab을 이용해서 Term Project를 할 수 있도록 환경을 준비해 줍니다. 터미널 사용이 원활하지 않은 상태이고, 데이터를 오랫동안 보관하기 어렵기 때문에 되도록 개인용 컴퓨터에서 [Anaconda](https://www.anaconda.com/)나 [JupyterLab](https://jupyterlab.readthedocs.io/en/stable/) 등을 설치하고 사용하시는 것이 편하지만, 개인용 컴퓨터 사양이 부족한 경우에는 Google CoLab을 사용할 수도 있습니다. 개인용 컴퓨터를 쓸 때는 Windows에서는 [WSL2](https://docs.microsoft.com/en-us/windows/wsl/install-win10)를 쓰시면 리눅스 커맨드라인 환경을 쓸 수 있습니다. Google Colab은 창을 닫은 후 90분 후, 또는 창을 처음 열고 12시간 후에 환경을 닫아버리기 때문에 매번 실습할 때마다 새로 셋업이 필요합니다. `Runtime->Run All`을 실행해서 전체를 한 번 실행하시면 준비됩니다. 모든 것이 잘 안 되고 뭔가 꼬인 느낌이 강하게 들 때는 `Runtime->Factory reset runtime`을 누르고 새로 시작하시면 될 수도 있습니다.\n"]},{"cell_type":"markdown","metadata":{"id":"zGYjheLcRJS0"},"source":["## Google Drive 마운트\n","\n","데이터 파일들을 불러오고 저장하기 위해서 Google Drive를 연결합니다. **이번엔 스크립트 파일을 편집해야 하기 때문에 Google Drive 마운트를 하지 않으면 vim등 콘솔에서 쓸 수 있는 에디터를 쓰셔야 합니다.** 화면에 나오는 주소를 눌러서 인증하고 인증 코드를 그 아래에 입력하세요. 그러면 `/contents/drive/MyDrive`에서 자신의 Drive 최상단 폴더가 보입니다. Google CoLab에서 접속하지 않고, 본인 컴퓨터 또는 연구실 워크스테이션 등 다른 컴퓨터를 쓰시는 경우에는 저장공간이 충분하므로 이 부분은 생략하고 지나갑니다."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0HgOxANiRHgl"},"outputs":[],"source":["from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"markdown","metadata":{"id":"ra3RB22DWEER"},"source":["## 실습용 Bioconda 환경 설치하기\n","Google CoLab에서 접속하지 않고, 본인 컴퓨터 또는 연구실 워크스테이션 등 다른 컴퓨터를 쓰시는 경우에는 따로 실습용 [Bioconda](https://bioconda.github.io/) 환경을 구축해서 이 노트북을 그 환경 안에서 띄우시면 됩니다.\n","\n","CoLab을 쓰시는 분들은 다음에 따라서 실습용 프로그램들을 쉽게 설치할 수 있도록 [Bioconda](https://bioconda.github.io/) 환경을 만듭니다. 실행하고 진행상황을 관찰하다가 에러로 보이는 것이 있으면 잘 읽어보고 해결합니다.\n","\n","파이썬 모듈들은 Colab쪽의 파이썬 인터프리터를 바꿀 수 없기 때문에 Conda에 설치하더라도 사용하기 어렵습니다. 파이썬 모듈은 `pip` 명령어로 설치하시면 Colab에서 불러와서 사용할 수 있습니다.\n","\n","환경 초기화에 관련된 스크립트는 [GitHub 프로젝트](https://github.com/hyeshik/colab-biolab)에 있습니다."]},{"cell_type":"code","execution_count":24,"metadata":{"id":"TTXmNi4hcap3"},"outputs":[{"name":"stdout","output_type":"stream","text":["fatal: destination path 'colab-biolab' already exists and is not an empty directory.\n","./\n","./root/\n","./root/.bashrc.biolab\n","./root/.bin.priority/\n","./root/.bin.priority/pip\n","./root/.bin.priority/pip2\n","./root/.bin.priority/pip3\n","./root/.condarc\n","./root/.profile\n","./root/.tmux.conf\n","./root/.vimrc\n","PREFIX=/root/conda\n","Unpacking payload ...\n","\n","Installing base environment...\n","\n","Preparing transaction: ...working... done\n","Executing transaction: ...working... done\n","installation finished.\n"]},{"ename":"ValueError","evalue":"not enough values to unpack (expected 2, got 1)","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","Cell \u001b[0;32mIn[24], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39msystem(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgit clone https://github.com/hyeshik/colab-biolab.git\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      2\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39msystem(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcd colab-biolab && bash tools/setup.sh\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m exec(\u001b[38;5;28mopen\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcolab-biolab/tools/activate_conda.py\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mread())\n","File \u001b[0;32m<string>:16\u001b[0m\n","\u001b[0;31mValueError\u001b[0m: not enough values to unpack (expected 2, got 1)"]}],"source":["!git clone https://github.com/hyeshik/colab-biolab.git\n","!cd colab-biolab && bash tools/setup.sh\n","exec(open('colab-biolab/tools/activate_conda.py').read())"]},{"cell_type":"markdown","metadata":{"id":"eDfJ0fCzsTTb"},"source":["Google CoLab에서 shell command를 실행할 때는 명령 앞에 !을 붙이면 실행할 수 있습니다."]},{"cell_type":"markdown","metadata":{"id":"J2LqGiP_N26s"},"source":["## 시퀀싱 데이터 다운로드\n","이제 실습 수업에서 필요한 시퀀싱 데이터를 다운로드 받아서 설치합시다. 드라이브 공간만 충분하면 데이터는 계속 저장돼 있기 때문에 학기 전체를 통해서 딱 한 번만 하시면 됩니다. Colab에서는 작업용 데이터를 Google Drive 폴더 아래의 `binfo-datapack1`에 저장합니다. 실습용 데이터는 `/content/drive/MyDrive/binfo1-datapack1`에 풀리게 됩니다.\n","\n","### (Option 1) Google CoLab에서 접속한 경우\n","1. [Google Drive 공유폴더](https://drive.google.com/drive/folders/1QBJednh-C7A17PFOlpZHBKyDmBvd6klh?usp=sharing)에 접속합니다.\n","1. 중간 제목줄의 \"binfo1-datapack1 ▾\"에서 ▾을 누릅니다.\n","1. Add shortcut to Drive를 누릅니다.\n","1. My Drive를 선택해서 ADD SHORTCUT 버튼을 누릅니다.\n","1. 다음 cell을 실행해서 링크가 제대로 들어왔는지 확인합니다. (바로 확인되지 않으면 3–5분 정도 기다렸다가 하면 됩니다.)\n","1. 제대로 확인되면 다음으로 넘어갑니다.\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"D24wKYriG8tI"},"outputs":[],"source":["!ls -al /content/drive/MyDrive/binfo1-datapack1/"]},{"cell_type":"markdown","metadata":{"id":"OZ0KkKrmG-M4"},"source":["### (Option 2) CoLab외의 다른 환경에서 접속한 경우\n","본인 컴퓨터나 연구실 워크스테이션 등에서 작업하는 경우에는 다음 명령어로 다운로드 받으시면 됩니다. 첫 번째에는 wget이 접속오류를 내는 경우가 많으니, 접속에러가 발생하면 다시 한 번 실행해 보세요. `/content/drive/Mydrive` 부분은 원하는 저장 경로로 바꿔서 지정하시면 됩니다."]},{"cell_type":"code","execution_count":3,"metadata":{"id":"kIN484e7NSxR"},"outputs":[{"name":"stdout","output_type":"stream","text":["--2024-05-08 14:52:18--  https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar\n","Resolving hyeshik.qbio.io (hyeshik.qbio.io)... 147.47.203.129\n","Connecting to hyeshik.qbio.io (hyeshik.qbio.io)|147.47.203.129|:443... connected.\n","WARNING: cannot verify hyeshik.qbio.io's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:\n","  Unable to locally verify the issuer's authority.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 6424668160 (6.0G) [application/octet-stream]\n","Saving to: ‘STDOUT’\n","\n","-                   100%[===================>]   5.98G  34.7MB/s    in 2m 51s  \n","\n","2024-05-08 14:55:09 (35.8 MB/s) - written to stdout [6424668160/6424668160]\n","\n"]}],"source":["!wget -O - --no-check-certificate https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar | tar -C /mnt/d/binfodata -xf -"]},{"cell_type":"markdown","metadata":{"id":"kiQ5y4dTD27M"},"source":["그리고, GENCODE Annotation 데이터도 미리 받아서 저장해 둡니다. 에러가 나면 다시 한 번 실행해 봅니다."]},{"cell_type":"code","execution_count":4,"metadata":{"id":"Efdd3x8hDMh4"},"outputs":[{"name":"stdout","output_type":"stream","text":["--2024-05-08 14:55:23--  http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz\n","Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.165\n","Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.165|:80... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 28359457 (27M) [application/x-gzip]\n","Saving to: ‘/mnt/d/binfodata/binfo1-datapack1/gencode.gtf.gz’\n","\n","/mnt/d/binfodata/bi 100%[===================>]  27.04M  71.8KB/s    in 3m 53s  \n","\n","2024-05-08 14:59:16 (119 KB/s) - ‘/mnt/d/binfodata/binfo1-datapack1/gencode.gtf.gz’ saved [28359457/28359457]\n","\n"]}],"source":["!wget --no-check-certificate -O /mnt/d/binfodata/binfo1-datapack1/gencode.gtf.gz http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz"]},{"cell_type":"markdown","metadata":{"id":"lnlM-S70C-tY"},"source":["### 파일 체크섬 확인\n","\n","받은 데이터의 MD5 체크섬을 확인합니다. 한참 걸립니다. 각 파일들의 원래 체크섬은 다음과 같습니다. 한 번 비교해 보세요. (나중에 자동으로 비교하고 싶으시면 `md5sum -c` 옵션을 사용하면 됩니다. 여기서는 MD5를 처음 써 보는 의미에서 직접 비교해 봅시다.)\n","\n","```\n","140aaf30bcb9276cc716f8699f04ddd6  CLIP-35L33G.bam\n","f1b3336ed7e2f97d562dcc71641251bd  CLIP-35L33G.bam.bai\n","328883a73d507eafbf5b60bd6b906201  RNA-control.bam\n","02073818e2f398a73c3b76e5169de1ca  RNA-control.bam.bai\n","b09550d09d6c2a4ce27f0226f426fdb1  RNA-siLin28a.bam\n","fef112c727244060ea62d3f2564a07f6  RNA-siLin28a.bam.bai\n","28bbd0c47d725669340c784f1b772c01  RNA-siLuc.bam\n","43590fdc4d81905c0432e0d1cb8cfd5b  RNA-siLuc.bam.bai\n","5c08a9297307bc83259e658c4474f0cc  RPF-siLin28a.bam\n","a1bb3e29be412dfd7fd8d16b1b1acc4c  RPF-siLin28a.bam.bai\n","f2eebf50943024d0116c9cd3e744c707  RPF-siLuc.bam\n","dc24f69e8f571fc8be30f28ce5b84fcd  RPF-siLuc.bam.bai\n","```"]},{"cell_type":"code","execution_count":7,"metadata":{"id":"g8lC_YJRBCPf"},"outputs":[{"name":"stdout","output_type":"stream","text":["md5sum: /mnt/d/binfodata/binfo1-datapack1/CLIP-35L33G.bam: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/CLIP-35L33G.bam.bai: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RNA-control.bam: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RNA-control.bam.bai: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RNA-siLin28a.bam: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RNA-siLin28a.bam.bai: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RNA-siLuc.bam: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RNA-siLuc.bam.bai: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RPF-siLin28a.bam: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RPF-siLin28a.bam.bai: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RPF-siLuc.bam: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/RPF-siLuc.bam.bai: no properly formatted MD5 checksum lines found\n","md5sum: /mnt/d/binfodata/binfo1-datapack1/gencode.gtf.gz: no properly formatted MD5 checksum lines found\n"]}],"source":["!md5sum -c /mnt/d/binfodata/binfo1-datapack1/*"]},{"cell_type":"code","execution_count":8,"metadata":{"id":"PVXzIs5FE3Q4"},"outputs":[{"name":"stdout","output_type":"stream","text":["total 6301792\n","drwxrwxrwx 1 root root       4096 May  8 14:55 .\n","drwxrwxrwx 1 root root       4096 May  8 14:52 ..\n","-rwxrwxrwx 1 root root 1370036258 May  6  2021 CLIP-35L33G.bam\n","-rwxrwxrwx 1 root root    3118336 May  6  2021 CLIP-35L33G.bam.bai\n","-rwxrwxrwx 1 root root 1003658801 May  6  2021 RNA-control.bam\n","-rwxrwxrwx 1 root root    2276104 May  6  2021 RNA-control.bam.bai\n","-rwxrwxrwx 1 root root 1260991122 May  6  2021 RNA-siLin28a.bam\n","-rwxrwxrwx 1 root root    2710744 May  6  2021 RNA-siLin28a.bam.bai\n","-rwxrwxrwx 1 root root  981684502 May  6  2021 RNA-siLuc.bam\n","-rwxrwxrwx 1 root root    2606104 May  6  2021 RNA-siLuc.bam.bai\n","-rwxrwxrwx 1 root root  737352902 May  6  2021 RPF-siLin28a.bam\n","-rwxrwxrwx 1 root root    2482768 May  6  2021 RPF-siLin28a.bam.bai\n","-rwxrwxrwx 1 root root 1055123479 May  6  2021 RPF-siLuc.bam\n","-rwxrwxrwx 1 root root    2605912 May  6  2021 RPF-siLuc.bam.bai\n","-rwxrwxrwx 1 root root   28359457 Mar  9  2021 gencode.gtf.gz\n"]}],"source":["!ls -al /mnt/d/binfodata/binfo*"]},{"cell_type":"markdown","metadata":{"id":"ecEA1PPQUlIK"},"source":["실습에 필요한 프로그램들 몇 가지를 추가로 설치합니다."]},{"cell_type":"code","execution_count":39,"metadata":{"id":"neteyWsWEKBy"},"outputs":[{"name":"stdout","output_type":"stream","text":["Channels:\n"," - conda-forge\n"," - bioconda\n"," - defaults\n","Platform: linux-64\n","Collecting package metadata (repodata.json): done\n","Solving environment: done\n","\n","# All requested packages already installed.\n","\n"]}],"source":["!conda install -y subread"]},{"cell_type":"markdown","metadata":{"id":"Oe_54MZrEi4y"},"source":["## Transcript count table 마련하기\n","\n","제공된 alignment와 annotation 파일들을 이용해서 transcript별 read count를 구해봅니다. 실제 연구에서는 multi-mapping 등을 고려해야 하지만, 여기서는 단순화해서 모두 무시합니다."]},{"cell_type":"markdown","metadata":{"id":"rrkL8Zl-s978"},"source":["먼저 작업 디렉토리를 만들고 데이터를 원본 디렉토리에서 복사해 옵니다. (이미 이 노트북을 실행한 적이 있으면 다시 할 필요는 없습니다.)"]},{"cell_type":"code","execution_count":9,"metadata":{"id":"ibYTwDaIon6b"},"outputs":[{"name":"stdout","output_type":"stream","text":["sending incremental file list\n","./\n","CLIP-35L33G.bam\n","  1,370,036,258 100%   63.35MB/s    0:00:20 (xfr#1, to-chk=12/14)\n","CLIP-35L33G.bam.bai\n","      3,118,336 100%    3.01MB/s    0:00:00 (xfr#2, to-chk=11/14)\n","RNA-control.bam\n","  1,003,658,801 100%   59.03MB/s    0:00:16 (xfr#3, to-chk=10/14)\n","RNA-control.bam.bai\n","      2,276,104 100%   14.97MB/s    0:00:00 (xfr#4, to-chk=9/14)\n","RNA-siLin28a.bam\n","  1,260,991,122 100%   60.19MB/s    0:00:19 (xfr#5, to-chk=8/14)\n","RNA-siLin28a.bam.bai\n","      2,710,744 100%    9.76MB/s    0:00:00 (xfr#6, to-chk=7/14)\n","RNA-siLuc.bam\n","    981,684,502 100%   59.33MB/s    0:00:15 (xfr#7, to-chk=6/14)\n","RNA-siLuc.bam.bai\n","      2,606,104 100%    2.64MB/s    0:00:00 (xfr#8, to-chk=5/14)\n","RPF-siLin28a.bam\n","    737,352,902 100%   61.38MB/s    0:00:11 (xfr#9, to-chk=4/14)\n","RPF-siLin28a.bam.bai\n","      2,482,768 100%    1.42MB/s    0:00:01 (xfr#10, to-chk=3/14)\n","RPF-siLuc.bam\n","  1,055,123,479 100%   63.08MB/s    0:00:15 (xfr#11, to-chk=2/14)\n","RPF-siLuc.bam.bai\n","      2,605,912 100%    5.67MB/s    0:00:00 (xfr#12, to-chk=1/14)\n","gencode.gtf.gz\n","     28,359,457 100%   31.30MB/s    0:00:00 (xfr#13, to-chk=0/14)\n","\n","sent 6,454,582,740 bytes  received 266 bytes  64,224,706.53 bytes/sec\n","total size is 6,453,006,489  speedup is 1.00\n"]}],"source":["!mkdir /mnt/d/binfodata/binfo1-work\n","!rsync -Pav /mnt/d/binfodata/binfo1-datapack1/ /mnt/d/binfodata/binfo1-work/"]},{"cell_type":"code","execution_count":40,"metadata":{"id":"WJEqpSd_rgYK"},"outputs":[{"name":"stdout","output_type":"stream","text":["/mnt/d/binfodata/binfo1-work\n"]}],"source":["%cd /mnt/d/binfodata/binfo1-work"]},{"cell_type":"markdown","metadata":{"id":"gudJrFYlsTun"},"source":["자! bam 파일에 담긴 aligned reads와 gtf 파일에 담긴 gene annotation으로 gene별 read count를 구합니다."]},{"cell_type":"code","execution_count":41,"metadata":{"id":"5OrnDxegFAip"},"outputs":[{"name":"stdout","output_type":"stream","text":["/bin/bash: line 1: featureCounts: command not found\n"]}],"source":["!featureCounts -a gencode.vM27.annotation.gtf -o read-counts.txt *.bam"]},{"cell_type":"markdown","metadata":{"id":"b2yJztZvGl9Q"},"source":["결과를 pandas로 읽어와 봅니다."]},{"cell_type":"code","execution_count":35,"metadata":{"id":"BMyNT3ARFAWi"},"outputs":[{"ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: 'read-counts.txt'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","Cell \u001b[0;32mIn[35], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m cnts \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mread-counts.txt\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, comment\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m, index_col\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m      3\u001b[0m cnts\u001b[38;5;241m.\u001b[39mhead()\n","File \u001b[0;32m~/junchae/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:948\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    936\u001b[0m     dialect,\n\u001b[1;32m    937\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    944\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    945\u001b[0m )\n\u001b[1;32m    946\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 948\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n","File \u001b[0;32m~/junchae/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:611\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    608\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    610\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 611\u001b[0m parser \u001b[38;5;241m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    613\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    614\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n","File \u001b[0;32m~/junchae/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1448\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1445\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1447\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1448\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_engine(f, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine)\n","File \u001b[0;32m~/junchae/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1705\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1703\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1704\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1705\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m get_handle(\n\u001b[1;32m   1706\u001b[0m     f,\n\u001b[1;32m   1707\u001b[0m     mode,\n\u001b[1;32m   1708\u001b[0m     encoding\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1709\u001b[0m     compression\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcompression\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1710\u001b[0m     memory_map\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmemory_map\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m),\n\u001b[1;32m   1711\u001b[0m     is_text\u001b[38;5;241m=\u001b[39mis_text,\n\u001b[1;32m   1712\u001b[0m     errors\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding_errors\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstrict\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[1;32m   1713\u001b[0m     storage_options\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstorage_options\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1714\u001b[0m )\n\u001b[1;32m   1715\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1716\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n","File \u001b[0;32m~/junchae/anaconda3/lib/python3.11/site-packages/pandas/io/common.py:863\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    858\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    859\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    860\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    861\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    862\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 863\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(\n\u001b[1;32m    864\u001b[0m             handle,\n\u001b[1;32m    865\u001b[0m             ioargs\u001b[38;5;241m.\u001b[39mmode,\n\u001b[1;32m    866\u001b[0m             encoding\u001b[38;5;241m=\u001b[39mioargs\u001b[38;5;241m.\u001b[39mencoding,\n\u001b[1;32m    867\u001b[0m             errors\u001b[38;5;241m=\u001b[39merrors,\n\u001b[1;32m    868\u001b[0m             newline\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    869\u001b[0m         )\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    871\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    872\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'read-counts.txt'"]}],"source":["import pandas as pd\n","cnts = pd.read_csv('read-counts.txt', sep='\\t', comment='#', index_col=0)\n","cnts.head()"]},{"cell_type":"markdown","metadata":{"id":"NyrUZ7WBPAmv"},"source":["이제 한 번 대충 그림을 그려 봅니다."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"YlSZiJgVPAZS"},"outputs":[],"source":["cnts['clip_enrichment'] = cnts['CLIP-35L33G.bam'] / cnts['RNA-control.bam']\n","cnts['rden_change'] = (cnts['RPF-siLin28a.bam'] / cnts['RNA-siLin28a.bam']) / (cnts['RPF-siLuc.bam'] / cnts['RNA-siLuc.bam'])\n","cnts.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"vU3ce_HEFALl"},"outputs":[],"source":["from matplotlib import pyplot as plt\n","import numpy as np\n","\n","fig, ax = plt.subplots(1, 1, figsize=(5, 5))\n","ax.scatter(np.log2(cnts['clip_enrichment']),\n","           np.log2(cnts['rden_change']))"]},{"cell_type":"markdown","metadata":{"id":"ksVjIL10mV_v"},"source":["이 그림에는 문제가 많이 있습니다. 논문의 그림 Figure 4D처럼 한 번 만들어 봅시다~"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"RZ-EV3s5X3u5"},"outputs":[],"source":["# 여기에 코드를 새로 작성해서 논문의 그림과 비슷하게 만들어 봅니다."]},{"cell_type":"markdown","metadata":{"id":"sf6EJJ60Qbxg"},"source":["## Protein localization 반영하기\n","\n","논문 그림(Figure 5B, S6A)에서는 protein localization을 반영해서 색으로 나타냈습니다. 이것을 한 번 구현해 봅시다. 각 gene product의 localization을 따로 구하려면 좀 귀찮기 때문에 미리 모아 두었습니다. UniProt에서 매핑된 데이터를 토대로 아주 명확한 것만 일부 정리한 것입니다. 이것도 마찬가지로 처음엔 에러가 나기도 하니까, 다시 실행해보면 거의 잘 됩니다~"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ivb1q7_f-gl4"},"outputs":[],"source":["import ssl\n","ssl._create_default_https_context = ssl._create_unverified_context\n","mouselocal = pd.read_csv('https://hyeshik.qbio.io/binfo/mouselocalization-20210507.txt', sep='\\t')\n","mouselocal.head()"]},{"cell_type":"markdown","metadata":{"id":"Q9vgSzeiQxDU"},"source":["이제 이 localization 데이터와 위에서 만든 scatter를 결합해서 논문 그림과 비슷하게 만들어 봅시다."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"aPPW73t4X3pp"},"outputs":[],"source":["# 여기에 새로 코드를 작성해서 논문 그림과 비슷하게 만들어 봅시다."]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.7"}},"nbformat":4,"nbformat_minor":0}
